src/Core/Content/Product/DataAbstractionLayer/ProductStreamUpdater.php

```diff
@@ -58,74 +60,67 @@ public function handle(EntityIndexingMessage $message): void
             return;
         }
 
-        $id = $message->getData();
-        if (!\is_string($id)) {
+        $streamId = $message->getData();
+        if (!\is_string($streamId)) {
             return;
         }
 
         $filter = $this->connection->fetchOne(
             'SELECT api_filter FROM product_stream WHERE invalid = 0 AND api_filter IS NOT NULL AND id = :id',
-            ['id' => Uuid::fromHexToBytes($id)]
+            ['id' => Uuid::fromHexToBytes($streamId)]
         );
         // if the filter is invalid
         if ($filter === false) {
             return;
         }
 
-        $insert = new MultiInsertQueryQueue($this->connection, 250, false, true);
-
         $version = Uuid::fromHexToBytes(Defaults::LIVE_VERSION);
 
         $filter = json_decode((string) $filter, true, 512, \JSON_THROW_ON_ERROR);
 
         $criteria = $this->getCriteria($filter);
-
         if ($criteria === null) {
             return;
         }
 
-        $criteria->setLimit(150);
         $considerInheritance = $message->getContext()->considerInheritance();
         $message->getContext()->setConsiderInheritance(true);
 
-        $iterator = new RepositoryIterator(
-            $this->repository,
-            $message->getContext(),
-            $criteria
-        );
-
-        $binary = Uuid::fromHexToBytes($id);
+        $binaryStreamId = Uuid::fromHexToBytes($streamId);
 
+        /** @var list<string> $ids */
         $ids = $this->connection->fetchFirstColumn(
             'SELECT LOWER(HEX(product_id)) FROM product_stream_mapping WHERE product_stream_id = :id',
-            ['id' => $binary],
+            ['id' => $binaryStreamId],
         );
 
-        RetryableTransaction::retryable($this->connection, function () use ($binary): void {
+        RetryableTransaction::retryable($this->connection, function () use ($binaryStreamId): void {
             $this->connection->executeStatement(
                 'DELETE FROM product_stream_mapping WHERE product_stream_id = :id',
-                ['id' => $binary],
+                ['id' => $binaryStreamId],
             );
         });
 
-        while ($matches = $iterator->fetchIds()) {
-            foreach ($matches as $id) {
-                if (!\is_string($id)) {
-                    continue;
-                }
-                $ids[] = $id;
-                $insert->addInsert('product_stream_mapping', [
-                    'product_id' => Uuid::fromHexToBytes($id),
-                    'product_version_id' => $version,
-                    'product_stream_id' => $binary,
-                ]);
-            }
+        /** @var list<string> $matches */
+        $matches = $this->repository->searchIds($criteria, $message->getContext())->getIds();
```

Comment:
I'm not sure if removing the iterator completely is really a good idea, as now we completely remove the upper bound 

My assumption would be that the iterator usage was getting slow as the default page size it uses is 50, which is definitely way too low for the usage here and adds a lot of overhead, did you try to set the limit on the criteria for the limit to smth like 10.000 or even 50.000? 
Is it still considerably slower then doing the query without limits?

Reply:
The RepositoryIterator.fetchIds is also using searchIds behind the scenes, so I don't get why it could be slower. In fact, it uses the product's auto increment, so it should be fast enough and safer than fetching everything at once. Maybe we can increase the Iterator chunk size so fewer queries are performed?